### PR TITLE
00887 d refactor test client error message handling

### LINKED ISSUE #887: Revisiting test-clients assertion usage to help HAPIClientValidator's error handling
**Summary**

In hedera services' performance regression tests, we frequently see some error messages in test-client's log files. Details please refer to issue #867 .

One reason for these client side log message issues is that test-clients were introduced as a testing tools for developer's usage. Now we are automating the regression test workflows, test-clients is to some extent treated as a product by itself.  Hence it may not be appropriate for test-clients to handle certain error conditions using Java assertions. 

**Possible resolution**
Revisit all the places where 'assertion' statements are used and check where the usage makes sense. If necessary, refactor it to use more appropriate error reporting;

### PR DESCRIPTION
**Pull request by ljianghedera**
**Related issue(s)**:
Closes #887 

**Summary of the change**:

1. changed the error logs to not throw stack trace in expected error scenarios;
2. Need check more test scenarios;

Here is one link to show the effect: https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1611036309063900. 

**External impacts**:
None.

**Applicable documentation**
- [ ] Javadoc
- [ ] HAPI protobuf comments
- [ ] README

### CODE REVIEW

File: .circleci/config.yml
Comment (Neeharika-Sompalli) [inline .circleci/config.yml:1408]: I believe the cron job changes cam be reverted

Comment (ljianghedera) [reply .circleci/config.yml:1408]: Yeah. I will keep this in mind. If there is no other places needs significant changes, I will post the cron job change to another issue to change/simplify config.yml that's in progress right now. There I need to redo the cron job time slots again to avoid conflicts of GCP resources.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/HapiApiSpec.java
Comment (qnswirlds) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/HapiApiSpec.java:330]: When we don't log an exception/throwable, rename it to `ignore`.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java
Comment (qnswirlds) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java:104]: Assign the `String.format` to a variable to remove duplicate.

**Review by qnswirlds — CHANGES_REQUESTED** [review]
`Assert.assertEquals` or `Assert.fail` will throw an `AssertionError` (or its extension). So I think it's better to use these methods than logging the error then throwing a generic `Exception`. If we want to just log the error without throwing an exception, then it is fine.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/HapiApiSpec.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/HapiApiSpec.java:330]: We still pack the exception message to let the finalizer to handle it.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java]: Done.

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM, apart from Quan's comment. Please merge master to this branch before merging.

File: .circleci/config.yml
Comment (Neeharika-Sompalli) [inline .circleci/config.yml:1154]: Looks like config.yml has extra changes, need to resolve conflicts with master branch.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiQueryCheckStateException.java
Comment (Neeharika-Sompalli) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiQueryCheckStateException.java:2]: copyright missing

File: test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiQueryPrecheckStateException.java
Comment (Neeharika-Sompalli) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiQueryPrecheckStateException.java:2]: same here

File: test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiTxnCheckStateException.java
Comment (Neeharika-Sompalli) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiTxnCheckStateException.java:2]: same here

File: test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiTxnPrecheckStateException.java
Comment (Neeharika-Sompalli) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiTxnPrecheckStateException.java:2]: same here

File: test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiQueryCheckStateException.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiQueryCheckStateException.java:2]: Added.

File: .circleci/config.yml
Comment (ljianghedera) [reply .circleci/config.yml:1154]: should be resolved.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiTxnPrecheckStateException.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/exceptions/HapiTxnPrecheckStateException.java:2]: Added.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java
Comment (qnswirlds) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java:107]: typo `actial` --> `actual`
Should use String.format as in other places to remove duplicate code

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/contract/HapiGetContractRecords.java
Comment (qnswirlds) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/contract/HapiGetContractRecords.java:215]: why not throw a newly created query exception?

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java
Comment (qnswirlds) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java:65]: Same here, throw a specific exception. I think originally `Assert.fail` was used for this purpose without going through the hassle of creating new specific exception. And it was better than throwing a generic exception here

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java
Comment (qnswirlds) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java:92]: Same here

**Review by qnswirlds — CHANGES_REQUESTED** [review]
I don't see why such changes will help solving the issue. Logging then rethrowing is exactly what Assert.fail or its kind does. And if in some places, we rethrow a _**generic**_ exception, it is worse. Since we want to avoid error messages from the those throws so not to mess up with the perf run, I would think we just log and not rethrowing.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java]: Done

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/contract/HapiGetContractRecords.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/queries/contract/HapiGetContractRecords.java]: Yeah, fixed.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java]: Fixed.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/queries/crypto/ExpectedTokenRel.java]: Here I will try to answer this and the another question above together. First, why only some places are touched. The reason is that the original Assert.fail() works well for many occasions. For example, when we are constructing a request (query or  transaction) to send to services, we know that we are trying to construct good (error free) requests to pass all the checks hopefully (unless we on purposely construct requests for negative functional test cases). So we don't expect any error to happen in those occasions, that's why I still keep original assertions untouched. For the situations when we wait for the results back, things are different. For the exact same good request, it may succeed this time, but fails in next run, and we should take these different outcome as normal in perf regression test, hence my changes are and should only focus on these places. Second,  replacing Assert.fail() is to avoid tons of stack trace in perf tests. Especially for long duration test, the log file can be excessive if not curbed. It's true we can still use Assert.fail(), but to avoid excessive stack trace, I need to do another catch in original exception's catch clause. That I think it's not necessary; Third, why re-throwing a new exception. The reason is if not, the already failed op would still go thru all following processing steps and we have to add logic to handle it. So re-throwing an exception will bypass these actions. Yeah, this issue takes much more time to decide which places to change than actually do the changes.

**Review by qnswirlds — APPROVED** [review]
To be honest, I am still not comfortable approving this PR. But I guess if we don't merge it, we will waste a lot of people time for nothing. Please provide a link to a passing performance test.

**Comment by ljianghedera** [comment]
> To be honest, I am still not comfortable approving this PR. But I guess if we don't merge it, we will waste a lot of people time for nothing. Please provide a link to a passing performance test.

This issue took more time to reach the stage as I explained. It took even more time to keep it updated with master branch's changes. Before we have better ideas to handle the error messages, I think we can do with it and at least keep the perf regression clean.

**Pull request merged**
